lxd/migrate_instance.go

```diff
@@ -152,7 +152,10 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
 	})
 	if err != nil {
 		l.Error("Failed migration on source", logger.Ctx{"err": err})
-		return fmt.Errorf("Failed migration on source: %w", err)
+
+		errMsg := fmt.Errorf("Failed migration on source: %w", err)
+		s.sendControl(errMsg)
```

Comment:
@boltmark what was the reason we couldnt cherry-pick this one by itself?

Reply:
@boltmark in what scenarios does this help with compared with https://github.com/canonical/lxd/pull/13894/files ?

Reply:
The reason we cannot use the cherry-pick by itself is because although it sends the error to the other side, this does not help if the other side does not read it. In the particular case for the `migration.stateful` check, the issue is twofold:
1. By not setting up the filesystem connection on the source before making this check, we cause the sink to fail before reading from the control connection. This means that even though we send over the error as per the cherry-pick, the sink never reads it in this case.
2. The first time the sink reads from the control connection, it does so expecting to receive a `MigrationHeader`. Therefore if the source fails before sending a header and instead sends an error, the sink will read the `MigrationControl` error message and (incorrectly) parse it into a `MigrationHeader`. The update to the `recv` function ensures we check for this case and return the error if one is found.

As for the particular scenarios this helps with compared to the other PR, this handles error checking for non-clustered servers, since in this case we cannot enforce that we always return the error for the source.

Reply:
Ah, sorry, I may have misread that. Do you mean why are we cherry-picking the added `sendControl` for both the source and the sink `Do` functions as opposed to just the source as in this code block?

Reply:
Yes please, this fix is in the client similar to what we discussed originally. Ta